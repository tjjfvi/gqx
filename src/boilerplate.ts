// @ts-nocheck
throw new Error("boilerplate.ts should not be imported")
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-unused-vars */

/* --- */

type Int = number;
type Float = number;
type String = string;
type ID = string;
type Boolean = boolean;

interface $<F, L> { f: F; l: L }
const $$ = <F, L>(f: F, l: L) => ({ f, l });

interface $$OperationId {
  typeProp: string;
  type: string;
  prop: string;
  inputTypes: { [x: string]: string };
}

const __wrap__ = Symbol();

type $$Wrap<X, Y> = X extends $<infer F, infer L> ? $<F, $$Wrap<L, Y>> : never;
type $$MapWrap<O, F> = {
  [K in keyof O | typeof __wrap__ | "$"]:
    K extends "$" ?
      <T extends $_>(x: T[]) => $<F, O extends { [__wrap__]: infer X } ? $$Wrap<X, T> : T>[] :
      K extends keyof O ?
        O[K] extends $_ ?
          $<F, O[K]> :
          $$MapWrap<O[K], F> :
        $<F, null>
}

const $$mapWrap = <O, F>(o: () => O, f: F): $$MapWrap<O, F> =>
  // @ts-ignore
  new Proxy(Object.create(null), {
    get: (t, k: string) => t[k] || (t[k] =
      k !== "$" ?
        o()[k].$ ?
          $$mapWrap(() => o()[k], f) :
          $$(f, o()[k]) :
        // @ts-ignore
        (a: any) => ("$" in o() ? o().$(a) : a).map((a: any) => $$(f, a))
    ),
  })

const $$reconstruct = <I extends $$OperationId>(id: I, input: $$Input.$<I>, props: $$Frag.$<I>[]) => {
  interface Subs { [k: string]: true | Subs }
  const subs: Subs = {};
  props.map(prop => populateSubs(prop, subs));
  const frag = genFrag(subs);
  const inputKeys = Object.keys(input);
  const inputDef = inputKeys.length ? `(${inputKeys.map(k =>
    `$${k}: ${id.inputTypes[k]}`
  ).join(", ")})` : "";
  const inputPass = inputKeys.length ? `(${inputKeys.map(k =>
    `${k}: $${k}`
  )})` : "";
  const request = `${id.typeProp}${inputDef} { ${id.prop}${inputPass}${frag} }`;
  return request;

  function genFrag(subs: Subs){
    return (
      Object.keys(subs).length ?
        `{ ${Object.entries(subs).map(([k, v]) => v === true ? k : k + " " + genFrag(v)).join(" ")} }` :
        ""
    );
  }

  function populateSubs(prop: $_, subs: Subs | true){
    if(subs === true)
      return;
    if("prop" in prop)
      subs[prop.prop] = true;
    else
      populateSubs(prop.l, subs[prop.f.prop] = (subs[prop.f.prop] || {}));
  }
}
