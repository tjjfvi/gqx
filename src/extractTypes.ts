import { DocumentNode, TypeNode } from "graphql"
import { Context } from "vm"
import { InputField, NamedType, Type, UnresolvedType } from "./type"

const baseScalars = ["ID", "String", "Int", "Float", "Boolean"]

export function extractTypes(documents: DocumentNode[]) {
  const types = new Map<string, NamedType>()

  for (const name of baseScalars) {
    types.set(name, {
      kind: "scalar",
      name,
      directives: [],
    })
  }

  const definitions = documents.flatMap(x => x.definitions)

  for (const typeDef of definitions) {
    if (typeDef.kind === "SchemaDefinition" || typeDef.kind === "SchemaExtension") {
      for (const operation of typeDef.operationTypes ?? []) {
        const type = getNamedType(operation.type.name.value)
        setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
        type.operationKey = operation.operation
      }
    }

    if (!("name" in typeDef) || !typeDef.name) continue

    const name = typeDef.name.value
    const type = getNamedType(name)

    if ("directives" in typeDef && typeDef.directives) {
      type.directives.push(...typeDef.directives)
    }

    if (typeDef.kind === "ScalarTypeDefinition" || typeDef.kind === "ScalarTypeExtension") {
      setKind(type, "scalar", {})
    } else if (
      typeDef.kind === "InputObjectTypeDefinition"
      || typeDef.kind === "InputObjectTypeExtension"
    ) {
      setKind(type, "input", { fields: [] })
      for (const fieldDef of typeDef.fields ?? []) {
        const key = fieldDef.name.value
        const fieldType = getType(fieldDef.type)
        type.fields.push({
          key,
          type: fieldType,
          directives: fieldDef.directives?.slice() ?? [],
        })
      }
    } else if (
      typeDef.kind === "ObjectTypeDefinition"
      || typeDef.kind === "ObjectTypeExtension"
      || typeDef.kind === "InterfaceTypeDefinition"
      || typeDef.kind === "InterfaceTypeExtension"
    ) {
      setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
      for (const fieldDef of typeDef.fields ?? []) {
        const key = fieldDef.name.value
        const fullType = getType(fieldDef.type)
        const namedType = getNamedTypeOfType(fullType)
        const name = `${type.name}$${key}`
        type.fields.push({
          str: JSON.stringify(name),
          name,
          base: type,
          key,
          namedType,
          fullType,
          directives: fieldDef.directives?.slice() ?? [],
          input: fieldDef.arguments?.map((arg): InputField => ({
            key: arg.name.value,
            type: getType(arg.type),
            directives: arg.directives?.slice() ?? [],
          })) ?? [],
        })
      }
      if ("interfaces" in typeDef) {
        for (const supertypeDef of typeDef.interfaces ?? []) {
          const supertype = getNamedType(supertypeDef.name.value)
          setKind(supertype, "object", { fields: [], subtypes: [], supertypes: [] })
          type.supertypes.push(supertype)
          supertype.subtypes.push(type)
        }
      }
    } else if (typeDef.kind === "UnionTypeDefinition" || typeDef.kind === "UnionTypeExtension") {
      setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
      for (const subtypeDef of typeDef.types ?? []) {
        const subtype = getNamedType(subtypeDef.name.value)
        setKind(subtype, "object", { fields: [], subtypes: [], supertypes: [] })
        type.subtypes.push(subtype)
        subtype.supertypes.push(type)
      }
    } else if (typeDef.kind === "EnumTypeDefinition" || typeDef.kind === "EnumTypeExtension") {
      setKind(type, "enum", { values: [] })
      for (const value of typeDef.values ?? []) {
        type.values.push({
          name: value.name.value,
          directives: value.directives?.slice() ?? [],
        })
      }
    } else {
      types.delete(name)
    }
  }

  return [...types.values()]

  function getNamedType(name: string): NamedType {
    const existing = types.get(name)
    if (existing !== undefined) {
      return existing
    }
    const value = { kind: "unresolved", name, directives: [] } as any as NamedType
    types.set(name, value)
    return value
  }

  function getType(node: TypeNode): Type {
    let nullable = true
    if (node.kind === "NonNullType") {
      node = node.type
      nullable = false
    }
    let type: Type
    if (node.kind === "NamedType") {
      type = getNamedType(node.name.value)
    } else {
      type = { kind: "array", inner: getType(node.type) }
    }
    if (nullable) {
      type = { kind: "nullable", inner: type }
    }
    return type
  }

  function setKind<K extends NamedType["kind"]>(
    type: NamedType | UnresolvedType,
    kind: K,
    init: Omit<Extract<NamedType, { kind: K }>, keyof UnresolvedType>,
  ): asserts type is Extract<NamedType, { kind: K }> {
    if ((type.kind === "unresolved") as boolean) {
      type.kind = kind
      Object.assign(type, init)
    }
    if (type.kind !== kind) throw new Error("Conflicting definitions for type " + type.name)
  }

  function getNamedTypeOfType(type: Type): NamedType {
    while (type.kind === "nullable" || type.kind === "array") {
      type = type.inner
    }
    return type
  }
}

export { Context }
