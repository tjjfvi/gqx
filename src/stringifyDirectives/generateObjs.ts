
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
