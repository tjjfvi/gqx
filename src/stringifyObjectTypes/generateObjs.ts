
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
