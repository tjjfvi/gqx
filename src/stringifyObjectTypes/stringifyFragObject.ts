
import { Obj, Prop } from ".";
import stringifyId from "./stringifyId";
import indent from "../indent";

export default (obj: Obj) => {
  const typeofType = `\n\ntype typeof$${obj.type} = typeof ${obj.type}`

  const fragCast = `const ${obj.type}$ = <F extends $$DeepArray<${obj.type}$>>(...frag: F) => [...frag];\n\n`

  const shallowObj = genObj(
    obj.shallowProps,
    p => stringifyId(p.id),
    indent`$: ${obj.type}$,\n`
  ) + " as const";

  const deepType = genObj(obj.deepProps, p =>
    `$$MapWrap<typeof$${p.type}, ${stringifyId(p.id)}>`,
  obj.shallowProps.length ? "" : indent`$: typeof ${obj.type}$,\n`);

  const deepObj = genObj(obj.deepProps, p =>
    `$$mapWrap(() => ${p.type}, ${stringifyId(p.id)})`,
  obj.shallowProps.length ? indent`..._${obj.type},\n` : indent`$: ${obj.type}$,\n`);

  if(!obj.deepProps.length)
    return fragCast + `export const ${obj.type} = ${shallowObj};` + typeofType;

  if(!obj.shallowProps.length)
    return fragCast + `export const ${obj.type}: ${deepType} = ${deepObj};` + typeofType;


  return fragCast + `
const _${obj.type} = ${shallowObj};

export const ${obj.type}: typeof _${obj.type} & ${deepType} = ${deepObj};
`.trim() + typeofType;
}

function genObj(props: Prop[], f: (p: Prop) => string, str = ""){
  if(!props.length)
    return "{}";
  return `{\n${str}${
    props.map(p =>
      indent`${p.id.prop}: ${f(p)},`
    ).join("\n")
  }\n}`;
}
