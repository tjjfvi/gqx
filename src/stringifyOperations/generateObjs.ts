
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
