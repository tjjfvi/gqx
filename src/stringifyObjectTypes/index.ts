
import { Context } from "..";
import stringifyFragEnums from "./stringifyFragEnums";
import stringifyFragResult from "./stringifyFragResult";
import { stringifyObjectTypeInfo } from "./stringifyObjectTypeInfo";
import stringifyFragObject from "./stringifyFragObject";
import generateObjs from "./generateObjs";

interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  shallow: boolean;
  wrap: (x: string) => string;
  wrapHKT: string;
}

interface Obj {
  type: string;
  shallowProps: Prop[];
  deepProps: Prop[];
}

export { Id, Prop, Obj, Context };

export default (ctx: Context) => {
  const [objs] = generateObjs(ctx);

  return [
    stringifyObjectTypeInfo(objs),
    stringifyFragEnums(objs),
    ...objs.map(o => stringifyFragObject(o)),
    ...objs.map(o => stringifyFragResult(o))
  ].join("\n\n");
}

