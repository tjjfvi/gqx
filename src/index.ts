
import {
  InputValueDefinitionNode,
  FieldDefinitionNode,
  EnumValueDefinitionNode,
  DirectiveNode,
} from "graphql";
import { stringifyEnumTypes } from "./stringifyEnumTypes";
import { stringifyInputTypes } from "./stringifyInputTypes";
import stringifyObjectTypes from "./stringifyObjectTypes";
import stringifyOperations from "./stringifyOperations";
import { stringifyBoilerplate } from "./stringifyBoilerplate";
import { groupDefinitions } from "./groupDefinitions";
import { stringifyTypeDirectives } from "./stringifyTypeDirectives";
import { parseSchema } from "./parseSchema";
import { stringifyScalarTypes } from "./stringifyScalarTypes";

interface Context {
  typeDirectives: Record<string, DirectiveNode[]>;
  inputTypes: { [k: string]: InputValueDefinitionNode[] };
  enumTypes: { [k: string]: EnumValueDefinitionNode[] };
  objectTypes: { [k: string]: { fields: FieldDefinitionNode[]; unions: string[]; implements: string[] } };
  operations: [string, string][];
  baseTypes: string[];
  scalarTypes: string[];
}

export default ({ schemaPath }: { schemaPath: string }) => {

  const ctx: Context = {
    operations: [],
    baseTypes: [],
    scalarTypes: [],
    enumTypes: {},
    inputTypes: {},
    objectTypes: {},
    typeDirectives: {},
  };

  const asts = parseSchema(schemaPath);

  groupDefinitions(ctx, asts.flatMap(x => x.definitions));

  const code = [
    "/* eslint-disable */",
    stringifyBoilerplate(),
    stringifyEnumTypes(ctx),
    stringifyInputTypes(ctx),
    stringifyObjectTypes(ctx),
    stringifyOperations(ctx),
    stringifyTypeDirectives(ctx),
    stringifyScalarTypes(ctx),
    "",
  ].join("\n\n").replace(/\n\n\n+/g, "\n\n");

  return code;

}

export { Context };
