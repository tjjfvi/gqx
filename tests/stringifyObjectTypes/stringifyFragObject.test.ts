
import { generateObjs } from "../../src/stringifyObjectTypes/generateObjs"
import { stringifyFragObject } from "../../src/stringifyObjectTypes/stringifyFragObject";
import { Obj } from "../../src/stringifyObjectTypes";

describe("stringifyObjectTypes.gql", () => {
  const [objs] = generateObjs(utils.getOperations(["stringifyObjectTypes"]));
  test.each(objs.map((o): [string, Obj] => [o.type, o]))("%s", (_, o) => {
    const code = stringifyFragObject(o);
    expect(code).toParse().toMatchSnapshot();
  })
})
