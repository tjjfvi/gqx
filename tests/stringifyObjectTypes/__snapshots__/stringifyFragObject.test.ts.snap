// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyObjectTypes.gql BinaryTree 1`] = `
"const BinaryTree$$ = <F extends $$DeepArray<BinaryTree$>>(...frag: F): F[number][] => frag;

const _BinaryTree = {
  $: BinaryTree$$,
  id: \\"BinaryTree$id\\",
  node: \\"BinaryTree$node\\",
} as const;

export const BinaryTree: typeof _BinaryTree & {
  left: $$MapWrap<typeof$BinaryTree, \\"BinaryTree$left\\">,
  right: $$MapWrap<typeof$BinaryTree, \\"BinaryTree$right\\">,
} = {
  ..._BinaryTree,
  left: $$mapWrap(() => BinaryTree, \\"BinaryTree$left\\"),
  right: $$mapWrap(() => BinaryTree, \\"BinaryTree$right\\"),
};

type typeof$BinaryTree = typeof BinaryTree"
`;

exports[`stringifyObjectTypes.gql DeepOnly 1`] = `
"const DeepOnly$$ = <F extends $$DeepArray<DeepOnly$>>(...frag: F): F[number][] => frag;

export const DeepOnly: {
  $: typeof DeepOnly$$,
  a: $$MapWrap<typeof$ShallowOnly, \\"DeepOnly$a\\">,
  b: $$MapWrap<typeof$ShallowOnly, \\"DeepOnly$b\\">,
} = {
  $: DeepOnly$$,
  a: $$mapWrap(() => ShallowOnly, \\"DeepOnly$a\\"),
  b: $$mapWrap(() => ShallowOnly, \\"DeepOnly$b\\"),
};

type typeof$DeepOnly = typeof DeepOnly"
`;

exports[`stringifyObjectTypes.gql Query 1`] = `
"const Query$$ = <F extends $$DeepArray<Query$>>(...frag: F): F[number][] => frag;

const _Query = {
  $: Query$$,
  oneDirective: \\"Query$oneDirective\\",
} as const;

export const Query: typeof _Query & {
  multDirectivesArgs: $$MapWrap<typeof$BinaryTree, \\"Query$multDirectivesArgs\\">,
  multDirectivesNoArgs: $$MapWrap<typeof$ShallowOnly, \\"Query$multDirectivesNoArgs\\">,
  multDirectivesNullArgs: $$MapWrap<typeof$DeepOnly, \\"Query$multDirectivesNullArgs\\">,
  noDirectives: $$MapWrap<typeof$Recursive, \\"Query$noDirectives\\">,
} = {
  ..._Query,
  multDirectivesArgs: $$mapWrap(() => BinaryTree, \\"Query$multDirectivesArgs\\"),
  multDirectivesNoArgs: $$mapWrap(() => ShallowOnly, \\"Query$multDirectivesNoArgs\\"),
  multDirectivesNullArgs: $$mapWrap(() => DeepOnly, \\"Query$multDirectivesNullArgs\\"),
  noDirectives: $$mapWrap(() => Recursive, \\"Query$noDirectives\\"),
};

type typeof$Query = typeof Query"
`;

exports[`stringifyObjectTypes.gql Recursive 1`] = `
"const Recursive$$ = <F extends $$DeepArray<Recursive$>>(...frag: F): F[number][] => frag;

const _Recursive = {
  $: Recursive$$,
  id: \\"Recursive$id\\",
} as const;

export const Recursive: typeof _Recursive & {
  next: $$MapWrap<typeof$Recursive, \\"Recursive$next\\">,
} = {
  ..._Recursive,
  next: $$mapWrap(() => Recursive, \\"Recursive$next\\"),
};

type typeof$Recursive = typeof Recursive"
`;

exports[`stringifyObjectTypes.gql ShallowOnly 1`] = `
"const ShallowOnly$$ = <F extends $$DeepArray<ShallowOnly$>>(...frag: F): F[number][] => frag;

export const ShallowOnly = {
  $: ShallowOnly$$,
  a: \\"ShallowOnly$a\\",
  b: \\"ShallowOnly$b\\",
} as const;

type typeof$ShallowOnly = typeof ShallowOnly"
`;

exports[`stringifyObjectTypes.gql WithId 1`] = `
"const WithId$$ = <F extends $$DeepArray<WithId$>>(...frag: F): F[number][] => frag;

export const WithId = {
  $: WithId$$,
  id: \\"WithId$id\\",
} as const;

type typeof$WithId = typeof WithId"
`;
